pkg/operator/controller/route-metrics/controller_test.go

```diff
@@ -0,0 +1,111 @@
+package routemetrics
```

Comment:
Please also test the Reconcile function when you add/update/delete a route and/or add/update/delete labels that would affect the number of routes per shard..

Reply:
I would also like to see an e2e test, which could be challenging but not impossible.  An e2e test would be a valuable contribution.  I didn't see any e2e tests for prometheus metric management in our repo, and having your example would be valuable.  I can help with this if needed.

Reply:
It would be great if you can point me to some unit tests written for the Reconcile function. The Reconcile function would be fetching the Ingress Controller object, Namespace List, and Route List from the apiserver. Can this be emulated in the unit test?

I would need your help with writing the e2e test. I'll take a look at the e2e tests in the repo for some inspiration.

Reply:
You don't need to test the Reconcile function directly.  What you could do is create a couple of private IngressControllers with labels.  Then create some routes with/without matching labels and check the metrics.  Then edit some routes or ingress controller and check the metrics.  Then delete some routes or ingress controller and check the metrics.

Reply:
> Yes, you can probably extract some functions from the body of the Reconcile and call those instead with pre-filled objects.
The `routeAdmitted` is already extracted out and a unit test is already created for it. The other parts in the reconcile function fetches the list of Namespaces and list of Routes from the cache.

Reply:
Check into pkg/operator/controller/ingress/metrics_test.go for unit tests.  

For e2e tests, maybe you could set up a pod that  allows you to query prometheus using curl, with a curl pod example in test/e2e/http_header_buffer_test.go.  I'll look into it.

Reply:
Though you may not need to do this with the go-client within the e2e test, I followed the instructions in https://docs.openshift.com/container-platform/4.11/monitoring/enabling-monitoring-for-user-defined-projects.html and was able to query the metric from the CLI.  Basically:
```
oc -n openshift-monitoring edit configmap cluster-monitoring-config
oc policy add-role-to-user monitoring-rules-view system:admin -n openshift-ingress
SECRET=`oc get secret -n openshift-user-workload-monitoring | grep  prometheus-user-workload-token | head -n 1 | awk '{print $1 }'`
export TOKEN=`echo $(oc get secret $SECRET -n openshift-user-workload-monitoring -o json | jq -r '.data.token') | base64 -d`
THANOS_QUERIER_HOST=`oc get route thanos-querier -n openshift-monitoring -o json | jq -r '.spec.host'`
curl -X GET -k  "https://$THANOS_QUERIER_HOST/api/v1/query?query=route_metrics_controller_routes_per_shard" -H "Authorization: Bearer $TOKEN"
```
which returns:
```
{"status":"success","data":{"resultType":"vector","result":[{"metric":{"__name__":"route_metrics_controller_routes_per_shard","container":"kube-rbac-proxy","endpoint":"metrics","instance":"10.128.0.31:9393","job":"metrics","name":"default","namespace":"openshift-ingress-operator","pod":"ingress-operator-7cb5db46f-d4h2m","prometheus":"openshift-monitoring/k8s","service":"metrics"},"value":[1663274395.426,"10"]}]}}
```

Reply:
I'll check this option for the e2e test.

Reply:
https://github.com/openshift/origin/blob/master/test/extended/router/metrics.go might be a helpful example.

Reply:
@Miciah, I was already looking at some examples in the origin repo. In many places a Prometheus client is used to fetch the metrics. It seems to be a clean and easy way of fetching metrics and verifying them. I have already tried it locally and it works fine. Should this approach be fine? Here is an example: https://github.com/openshift/origin/blob/master/pkg/monitor/alerts.go#L79-L118

Reply:
I'm not familiar with the code at <https://github.com/openshift/origin/blob/master/pkg/monitor/alerts.go#L79-L118>, but it looks fine, so if it works for the E2E that you're writing, then it should be fine to follow that example.